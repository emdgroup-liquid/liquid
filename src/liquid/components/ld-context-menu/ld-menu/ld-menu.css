:host {
  display: block;
}

.ld-menu {
  background: var(--ld-col-wht);
  border-radius: var(--ld-br-l);
  box-shadow: var(--ld-shadow-stacked);
  box-sizing: border-box;
  list-style: none;
  margin: 0;
  padding: var(--ld-sp-12);
}
