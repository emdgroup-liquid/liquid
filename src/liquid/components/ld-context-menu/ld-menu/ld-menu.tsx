import {
  Component,
  Element,
  h,
  Host,
  Method,
  Prop,
  State,
  Watch,
} from '@stencil/core'
import { isElement, isMenuItem, isSlot } from '../../../utils/type-checking'

const getMenuItemOrNestedMenuItems = (node: Node) => {
  if (!isElement(node)) {
    return []
  }

  if (isMenuItem(node)) {
    return [node]
  }

  const items: HTMLLdMenuitemElement[] = []

  if (isSlot(node)) {
    node
      .assignedNodes()
      .forEach((node) => items.push(...getMenuItemOrNestedMenuItems(node)))
    return items
  }

  node.childNodes.forEach((node) =>
    items.push(...getMenuItemOrNestedMenuItems(node))
  )

  return items
}

const isPrintableCharacter = (key: string) =>
  key.length === 1 && key.match(/\S/)

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part list - `ul` element wrapping the default slot
 */
@Component({
  tag: 'ld-menu',
  styleUrl: 'ld-menu.css',
  shadow: true,
})
export class LdMenu {
  @Element() el: HTMLLdMenuElement

  /** Size of the context menu. */
  @Prop() size?: 'sm' | 'lg'

  @State() initialized = false

  private initMenuItems = (element: Node, initial = false) => {
    if (!isElement(element)) {
      return
    }

    if (isMenuItem(element)) {
      element.size = this.size

      if (!initial) {
        return
      }

      element.ldTabindex = this.initialized ? -1 : 0

      if (!this.initialized) {
        this.initialized = true
      }

      return
    }

    if (isSlot(element)) {
      element.assignedNodes().forEach((node) => this.initMenuItems(node))
      return
    }

    element.childNodes.forEach((node) => this.initMenuItems(node))
  }

  /** Get the first menu item inside this menu. */
  @Method()
  async getFirstMenuItem(): Promise<HTMLLdMenuitemElement> {
    return this.getAllMenuItems()[0]
  }

  private getAllMenuItems = () => {
    const items: HTMLLdMenuitemElement[] = []

    this.el
      .querySelectorAll('slot, ld-menuitem')
      .forEach((node) => items.push(...getMenuItemOrNestedMenuItems(node)))

    return items
  }

  private focusFirst = (target: HTMLLdMenuitemElement) => {
    const allMenuItems = this.getAllMenuItems()
    const [first] = allMenuItems

    target.ldTabindex = -1
    first.ldTabindex = 0
    first.focusInner()
  }

  private focusLast = (target: HTMLLdMenuitemElement) => {
    const allMenuItems = this.getAllMenuItems()
    const last = allMenuItems[allMenuItems.length - 1]

    target.ldTabindex = -1
    last.ldTabindex = 0
    last.focusInner()
  }

  private focusNext = (target: HTMLLdMenuitemElement) => {
    const allMenuItems = this.getAllMenuItems()
    const index = allMenuItems.indexOf(target)
    const next =
      allMenuItems.length > index + 1
        ? allMenuItems[index + 1]
        : allMenuItems[0]

    target.ldTabindex = -1
    next.ldTabindex = 0
    next.focusInner()
  }

  private focusPrev = (target: HTMLLdMenuitemElement) => {
    const allMenuItems = this.getAllMenuItems()
    const index = allMenuItems.indexOf(target)
    const prev =
      index === 0
        ? allMenuItems[allMenuItems.length - 1]
        : allMenuItems[index - 1]

    target.ldTabindex = -1
    prev.ldTabindex = 0
    prev.focusInner()
  }

  private focusByFirstCharacter = (event: KeyboardEvent) => {
    if (!isPrintableCharacter(event.key)) {
      return
    }

    const target = event.target as HTMLLdMenuitemElement
    const allMenuItems = this.getAllMenuItems()
    const currentIndex = allMenuItems.indexOf(target)
    const itemMatchesKey = (item: HTMLLdMenuitemElement) =>
      item.textContent.trim()[0].toLocaleLowerCase() ===
      event.key.toLocaleLowerCase()
    let matchingItem = allMenuItems.find((item, index) => {
      if (index <= currentIndex) {
        return
      }

      return itemMatchesKey(item)
    })

    if (!matchingItem) {
      matchingItem = allMenuItems.find(itemMatchesKey)
    }

    if (!matchingItem || matchingItem === target) {
      return
    }

    target.ldTabindex = -1
    matchingItem.ldTabindex = 0
    matchingItem.focusInner()
  }

  private handleKeyDown = (event: KeyboardEvent) => {
    const target = event.target as HTMLLdMenuitemElement

    switch (event.key) {
      case 'ArrowUp':
        event.preventDefault()
        this.focusPrev(target)
        break
      case 'ArrowDown':
        event.preventDefault()
        this.focusNext(target)
        break
      case 'Home':
        event.preventDefault()
        this.focusFirst(target)
        break
      case 'End':
        event.preventDefault()
        this.focusLast(target)
        break
      default:
        this.focusByFirstCharacter(event)
    }
  }

  @Watch('size')
  handleSizeChange() {
    this.updateMenuItems()
  }

  private updateMenuItems = (initial = false) => {
    this.el
      .querySelectorAll('slot, ld-menuitem')
      .forEach((element) => this.initMenuItems(element, initial))
  }

  componentWillLoad() {
    this.updateMenuItems(true)
  }

  render() {
    return (
      <Host onKeyDown={this.handleKeyDown}>
        <ul class="ld-menu" part="list" role="menu">
          <slot />
        </ul>
      </Host>
    )
  }
}
