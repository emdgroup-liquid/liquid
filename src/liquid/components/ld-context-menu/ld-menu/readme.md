---
eleventyNavigation:
  key: Menu
  parent: Context Menu
layout: layout.njk
title: Menu
permalink: components/ld-context-menu/ld-menu/
---

<link rel="stylesheet" href="css_components/ld-menu.css">
<link rel="stylesheet" href="css_components/ld-menuitem.css">
<link rel="stylesheet" href="css_components/ld-menuitem-group.css">
<link rel="stylesheet" href="css_components/ld-icon.css">
<link rel="stylesheet" href="css_components/ld-button.css">

# ld-menu

The `ld-menu` component is used internally for the `ld-menu`. It is the visual part of the actual menu without the floating/tooltip functionality.

## Examples

### Basic

{% example %}
<ld-menu>
  <ld-menuitem>Menu item 1</ld-menuitem>
  <ld-menuitem>Menu item 2</ld-menuitem>
  <ld-menuitem>Menu item 3</ld-menuitem>
</ld-menu>

<!-- React component -->

<LdMenu>
  <LdMenuitem>Menu item 1</LdMenuitem>
  <LdMenuitem>Menu item 2</LdMenuitem>
  <LdMenuitem>Menu item 3</LdMenuitem>
</LdMenu>

<!-- CSS component -->

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 1
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 2
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 2
    </button>
  </li>
</ul>
{% endexample %}

### Size

{% example %}
<ld-menu size="sm">
  <ld-menuitem>Menu item 1</ld-menuitem>
  <ld-menuitem>Menu item 2</ld-menuitem>
  <ld-menuitem>Menu item 3</ld-menuitem>
</ld-menu>

<ld-menu>
  <ld-menuitem>Menu item 1</ld-menuitem>
  <ld-menuitem>Menu item 2</ld-menuitem>
  <ld-menuitem>Menu item 3</ld-menuitem>
</ld-menu>

<ld-menu size="lg">
  <ld-menuitem>Menu item 1</ld-menuitem>
  <ld-menuitem>Menu item 2</ld-menuitem>
  <ld-menuitem>Menu item 3</ld-menuitem>
</ld-menu>

<!-- React component -->

<LdMenu size="sm">
  <LdMenuitem>Menu item 1</LdMenuitem>
  <LdMenuitem>Menu item 2</LdMenuitem>
  <LdMenuitem>Menu item 3</LdMenuitem>
</LdMenu>

<LdMenu>
  <LdMenuitem>Menu item 1</LdMenuitem>
  <LdMenuitem>Menu item 2</LdMenuitem>
  <LdMenuitem>Menu item 3</LdMenuitem>
</LdMenu>

<LdMenu size="lg">
  <LdMenuitem>Menu item 1</LdMenuitem>
  <LdMenuitem>Menu item 2</LdMenuitem>
  <LdMenuitem>Menu item 3</LdMenuitem>
</LdMenu>

<!-- CSS component -->

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost ld-button--sm">
      Menu item 1
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost ld-button--sm">
      Menu item 2
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost ld-button--sm">
      Menu item 2
    </button>
  </li>
</ul>

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 1
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 2
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 2
    </button>
  </li>
</ul>

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost ld-button--lg">
      Menu item 1
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost ld-button--lg">
      Menu item 2
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost ld-button--lg">
      Menu item 2
    </button>
  </li>
</ul>
{% endexample %}

### Item Groups

{% example %}
<ld-menu>
  <ld-menuitem-group>
    <ld-menuitem>Menu item 1</ld-menuitem>
    <ld-menuitem>Menu item 2</ld-menuitem>
    <ld-menuitem>Menu item 3</ld-menuitem>
  </ld-menuitem-group>
  <ld-menuitem-group>
    <ld-menuitem mode="highlight">
      <ld-icon name="repost"></ld-icon>
      Reset
    </ld-menuitem>
    <ld-menuitem mode="danger">
      <ld-icon name="bin"></ld-icon>
      Delete
    </ld-menuitem>
  </ld-menuitem-group>
</ld-menu>

<!-- React component -->

<LdMenu>
  <LdMenuitemGroup>
    <LdMenuitem>Menu item 1</LdMenuitem>
    <LdMenuitem>Menu item 2</LdMenuitem>
    <LdMenuitem>Menu item 3</LdMenuitem>
  </LdMenuitemGroup>
  <LdMenuitemGroup>
    <LdMenuitem mode="highlight">
      <LdIcon name="repost" />
      Reset
    </LdMenuitem>
    <LdMenuitem mode="danger">
      <LdIcon name="bin" />
      Delete
    </LdMenuitem>
  </LdMenuitemGroup>
</LdMenu>

<!-- CSS component -->

<ul class="ld-menu" role="menu">
  <li class="ld-menuitem-group" role="none">
    <ul class="ld-menuitem-group__list" role="group">
      <li class="ld-menuitem"  role="menuitemradio">
        <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
          Menu item 1
        </button>
      </li>
      <li class="ld-menuitem"  role="menuitemradio">
        <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
          Menu item 2
        </button>
      </li>
      <li class="ld-menuitem"  role="menuitemradio">
        <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
          Menu item 2
        </button>
      </li>
    </ul>
  </li>
  <li class="ld-menuitem-group" role="none">
    <ul class="ld-menuitem-group__list" role="group">
      <li class="ld-menuitem"  role="menuitemradio">
        <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost">
          <svg class="ld-icon" role="presentation" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M18.7071 15.7071a.9972.9972 0 0 1-.3243.217 1.0001 1.0001 0 0 1-1.0899-.217l-3-3c-.3905-.3905-.3905-1.0237 0-1.4142.3905-.3905 1.0237-.3905 1.4142 0L17 12.5858V8h-7c-.5523 0-1-.4477-1-1s.4477-1 1-1h8c.5523 0 1 .4477 1 1v5.5858l1.2929-1.2929c.3905-.3905 1.0237-.3905 1.4142 0 .3905.3905.3905 1.0237 0 1.4142l-3 3ZM5.2929 8.2929a.9971.9971 0 0 1 .3243-.217 1.0003 1.0003 0 0 1 1.09.217l3 3c.3904.3905.3904 1.0237 0 1.4142-.3906.3905-1.0238.3905-1.4143 0L7 11.4142V16h7c.5523 0 1 .4477 1 1s-.4477 1-1 1H6c-.5523 0-1-.4477-1-1v-5.5858l-1.2929 1.2929c-.3905.3905-1.0237.3905-1.4142 0-.3905-.3905-.3905-1.0237 0-1.4142l3-3Z" fill="currentcolor"/></svg>
          Reset
        </button>
      </li>
      <li class="ld-menuitem"  role="menuitemradio">
        <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--danger-ghost">
          <svg class="ld-icon" role="presentation" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M16 3.6663h2.6669c.7369 0 1.3331.7463 1.3331 1.6663V6.5a.5.5 0 0 1-.5.5h-15a.5.5 0 0 1-.5-.5V5.3326c0-.92.597-1.6663 1.3339-1.6663H8A2.6517 2.6517 0 0 1 10.4618 2h3.0776c1.0838 0 2.0583.66 2.4606 1.6663ZM5.55 8a.5.5 0 0 0-.4978.5474L6 18.5S6 22 12 22s6-3.5 6-3.5l.9479-9.9526A.5.5 0 0 0 18.4501 8H5.5499ZM11 17.5v-6a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5v6a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5Zm4-6v6a.5.5 0 0 1-.5.5h-1a.5.5 0 0 1-.5-.5v-6a.5.5 0 0 1 .5-.5h1a.5.5 0 0 1 .5.5Z" fill="currentcolor"/></svg>
          Delete
        </button>
      </li>
    </ul>
  </li>
</ul>
{% endexample %}

### With custom width

{% example %}
<ld-menu style="width:18rem">
  <ld-menuitem>Menu item 1</ld-menuitem>
  <ld-menuitem>Menu item 2</ld-menuitem>
  <ld-menuitem>Menu item 3</ld-menuitem>
</ld-menu>

<!-- React component -->

<LdMenu style={ { width: '18rem' } }>
  <LdMenuitem>Menu item 1</LdMenuitem>
  <LdMenuitem>Menu item 2</LdMenuitem>
  <LdMenuitem>Menu item 3</LdMenuitem>
</LdMenu>

<!-- CSS component -->

<ul class="ld-menu" role="menu" style="width:18rem">
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 1
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 2
    </button>
  </li>
  <li class="ld-menuitem"  role="menuitemradio">
    <button class="ld-menuitem__button ld-button ld-button--justify-start ld-button--ghost ld-button--neutral-ghost">
      Menu item 2
    </button>
  </li>
</ul>
{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description               | Type           | Default     |
| -------- | --------- | ------------------------- | -------------- | ----------- |
| `size`   | `size`    | Size of the context menu. | `"lg" \| "sm"` | `undefined` |


## Methods

### `getFirstMenuItem() => Promise<HTMLLdMenuitemElement>`

Get the first menu item inside this menu.

#### Returns

Type: `Promise<HTMLLdMenuitemElement>`




## Shadow Parts

| Part     | Description |
| -------- | ----------- |
| `"list"` |             |


## Dependencies

### Used by

 - [ld-context-menu](..)

### Graph
```mermaid
graph TD;
  ld-context-menu --> ld-menu
  style ld-menu fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
