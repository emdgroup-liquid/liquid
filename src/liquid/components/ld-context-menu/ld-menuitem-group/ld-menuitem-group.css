:host {
  display: block;
}

.ld-menuitem-group__list {
  list-style: none;
  margin: 0;
  padding: 0;
}

:host(:not(:first-of-type)),
.ld-menuitem-group:not(:first-of-type) {
  border-top: solid var(--ld-sp-2) var(--ld-col-neutral-050);
  margin-top: var(--ld-sp-12);
  padding-top: var(--ld-sp-12);
}
