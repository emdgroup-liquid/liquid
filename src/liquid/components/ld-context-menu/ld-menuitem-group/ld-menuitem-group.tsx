import { Component, Host, h, Prop } from '@stencil/core'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part list - `ul` element wrapping the default slot
 * @part listitem - `li` element wrapping the `ul` element
 */
@Component({
  tag: 'ld-menuitem-group',
  styleUrl: 'ld-menuitem-group.css',
  shadow: true,
})
export class LdMenuitemGroup {
  /** Label for the menu item group. */
  @Prop() ariaLabel: HTMLUListElement['ariaLabel']

  render() {
    return (
      <Host>
        <li class="ld-menuitem-group" part="listitem" role="none">
          <ul
            class="ld-menuitem-group__list"
            aria-label={this.ariaLabel}
            role="group"
          >
            <slot></slot>
          </ul>
        </li>
      </Host>
    )
  }
}
