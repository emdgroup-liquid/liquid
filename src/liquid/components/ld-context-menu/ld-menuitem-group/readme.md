---
eleventyNavigation:
  key: Menuitem Group
  parent: Context Menu
layout: layout.njk
title: Menuitem Group
permalink: components/ld-context-menu/ld-menuitem-group/
---

# ld-menuitem-group

The `ld-menuitem-group` component is a subcomponent for `ld-context-menu`.

Please refer to the [`ld-context-menu` documentation](components/ld-context-menu) for usage examples.

---

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
