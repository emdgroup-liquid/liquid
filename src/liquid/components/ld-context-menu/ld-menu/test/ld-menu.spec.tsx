import { h } from "@stencil/core";
import { newSpecPage } from "@stencil/core/testing";
import { LdMenuitemGroup } from "../../ld-menuitem-group/ld-menuitem-group";
import { LdMenuitem } from "../../ld-menuitem/ld-menuitem";
import { LdMenu } from "../ld-menu";
import "../../../../utils/mutationObserver";

describe("ld-menu", () => {
  it("renders", async () => {
    const page = await newSpecPage({
      components: [LdMenu],
      template: () => (
        <ld-menu>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-menu>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it("forwards size prop to menu items", async () => {
    const page = await newSpecPage({
      components: [LdMenu, LdMenuitem, LdMenuitemGroup],
      template: () => (
        <ld-menu size="sm">
          <ld-menuitem>Menu item 1</ld-menuitem>
          <ld-menuitem-group>
            <ld-menuitem>Menu item 2</ld-menuitem>
            <ld-menuitem>Menu item 3</ld-menuitem>
          </ld-menuitem-group>
        </ld-menu>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it("forwards size prop changes to menu items", async () => {
    const page = await newSpecPage({
      components: [LdMenu, LdMenuitem, LdMenuitemGroup],
      template: () => (
        <ld-menu size="sm">
          <ld-menuitem>Menu item 1</ld-menuitem>
          <ld-menuitem-group>
            <ld-menuitem>Menu item 2</ld-menuitem>
            <ld-menuitem>Menu item 3</ld-menuitem>
          </ld-menuitem-group>
        </ld-menu>
      ),
    });

    page.root.size = "lg";
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  describe("keyboard navigation", () => {
    let item1: HTMLLdMenuitemElement,
      item2: HTMLLdMenuitemElement,
      item3: HTMLLdMenuitemElement,
      item4: HTMLLdMenuitemElement;

    beforeEach(async () => {
      const page = await newSpecPage({
        components: [LdMenu, LdMenuitem, LdMenuitemGroup],
        template: () => (
          <ld-menu size="sm">
            <ld-menuitem>First menu item</ld-menuitem>
            <ld-menuitem>Second menu item</ld-menuitem>
            <ld-menuitem-group>
              <ld-menuitem>Third menu item</ld-menuitem>
              <ld-menuitem>Fourth menu item</ld-menuitem>
            </ld-menuitem-group>
          </ld-menu>
        ),
      });

      const items = Array.from(page.root.querySelectorAll("ld-menuitem"));
      [item1, item2, item3, item4] = items;

      items.forEach((item) => {
        Object.defineProperty(item, "focusInner", {
          value: jest.fn(),
          configurable: true,
          writable: true,
        });
      });
    });

    it("focuses items with a matching first character", async () => {
      item1.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "t",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item2.focusInner).not.toHaveBeenCalled();
      expect(item3.focusInner).toHaveBeenCalled();

      jest.advanceTimersByTime(500);
      item3.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "f",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item4.focusInner).toHaveBeenCalled();

      jest.advanceTimersByTime(500);
      item4.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "f",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item1.focusInner).toHaveBeenCalled();

      jest.advanceTimersByTime(500);
      item4.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "f",
          bubbles: true,
          composed: true,
        }),
      );
      item4.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "i",
          bubbles: true,
          composed: true,
        }),
      );
      item4.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "s",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item1.focusInner).toHaveBeenCalledTimes(3);
      expect(item2.focusInner).toHaveBeenCalledTimes(0);
      expect(item3.focusInner).toHaveBeenCalledTimes(1);
      expect(item4.focusInner).toHaveBeenCalledTimes(2);
    });

    it("focuses last item on End key", async () => {
      item1.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "End",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item4.focusInner).toHaveBeenCalled();
    });

    it("focuses first item on Home key", async () => {
      item3.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "Home",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item1.focusInner).toHaveBeenCalled();
    });

    it("focuses last item on ArrowDown + meta key", async () => {
      item1.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "ArrowDown",
          bubbles: true,
          composed: true,
          metaKey: true,
        }),
      );

      expect(item4.focusInner).toHaveBeenCalled();
    });

    it("focuses first item on ArrowUp + meta key", async () => {
      item3.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "ArrowUp",
          bubbles: true,
          composed: true,
          metaKey: true,
        }),
      );

      expect(item1.focusInner).toHaveBeenCalled();
    });

    it("focuses next item on ArrowDown key", async () => {
      item1.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "ArrowDown",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item2.focusInner).toHaveBeenCalled();

      item2.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "ArrowDown",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item3.focusInner).toHaveBeenCalled();
    });

    it("focuses previous item on ArrowUp key", async () => {
      item3.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "ArrowUp",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item2.focusInner).toHaveBeenCalled();

      item2.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "ArrowUp",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item1.focusInner).toHaveBeenCalled();
    });

    it("focuses first item on ArrowDown key, if last item is focused", async () => {
      item4.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "ArrowDown",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item1.focusInner).toHaveBeenCalled();
    });

    it("focuses last item on ArrowUp key, if first item is focused", async () => {
      item1.dispatchEvent(
        new KeyboardEvent("keydown", {
          key: "ArrowUp",
          bubbles: true,
          composed: true,
        }),
      );

      expect(item4.focusInner).toHaveBeenCalled();
    });
  });

  it("does not throw when disconnecting before hydration", () => {
    const component = new LdMenu();
    component.disconnectedCallback();
  });

  it("does not throw when disconnecting after typeAheadHandler initialization", () => {
    const component = new LdMenu();
    component.componentWillLoad();
    expect(component.typeAheadHandler).toBeTruthy();
    component.disconnectedCallback();
  });
});
