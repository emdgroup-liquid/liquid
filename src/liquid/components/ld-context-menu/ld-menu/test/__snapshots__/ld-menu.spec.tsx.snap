// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-menu forwards size prop changes to menu items 1`] = `
<ld-menu>
  <mock:shadow-root>
    <ul class="ld-menu" part="list" role="menu">
      <slot></slot>
    </ul>
  </mock:shadow-root>
  <ld-menuitem>
    <mock:shadow-root>
      <li class="ld-menuitem" role="menuitem">
        <ld-button class="ld-menuitem__button" justifycontent="start" ldtabindex="0" mode="neutral-ghost" size="lg" type="button">
          <slot></slot>
        </ld-button>
      </li>
    </mock:shadow-root>
    Menu item 1
  </ld-menuitem>
  <ld-menuitem-group>
    <mock:shadow-root>
      <li class="ld-menuitem-group" role="none">
        <ul class="ld-menuitem-group__list" role="group">
          <slot></slot>
        </ul>
      </li>
    </mock:shadow-root>
    <ld-menuitem>
      <mock:shadow-root>
        <li class="ld-menuitem" role="menuitem">
          <ld-button class="ld-menuitem__button" justifycontent="start" ldtabindex="-1" mode="neutral-ghost" size="lg" type="button">
            <slot></slot>
          </ld-button>
        </li>
      </mock:shadow-root>
      Menu item 2
    </ld-menuitem>
    <ld-menuitem>
      <mock:shadow-root>
        <li class="ld-menuitem" role="menuitem">
          <ld-button class="ld-menuitem__button" justifycontent="start" ldtabindex="-1" mode="neutral-ghost" size="lg" type="button">
            <slot></slot>
          </ld-button>
        </li>
      </mock:shadow-root>
      Menu item 3
    </ld-menuitem>
  </ld-menuitem-group>
</ld-menu>
`;

exports[`ld-menu forwards size prop to menu items 1`] = `
<ld-menu>
  <mock:shadow-root>
    <ul class="ld-menu" part="list" role="menu">
      <slot></slot>
    </ul>
  </mock:shadow-root>
  <ld-menuitem>
    <mock:shadow-root>
      <li class="ld-menuitem" role="menuitem">
        <ld-button class="ld-menuitem__button" justifycontent="start" ldtabindex="0" mode="neutral-ghost" size="sm" type="button">
          <slot></slot>
        </ld-button>
      </li>
    </mock:shadow-root>
    Menu item 1
  </ld-menuitem>
  <ld-menuitem-group>
    <mock:shadow-root>
      <li class="ld-menuitem-group" role="none">
        <ul class="ld-menuitem-group__list" role="group">
          <slot></slot>
        </ul>
      </li>
    </mock:shadow-root>
    <ld-menuitem>
      <mock:shadow-root>
        <li class="ld-menuitem" role="menuitem">
          <ld-button class="ld-menuitem__button" justifycontent="start" ldtabindex="-1" mode="neutral-ghost" size="sm" type="button">
            <slot></slot>
          </ld-button>
        </li>
      </mock:shadow-root>
      Menu item 2
    </ld-menuitem>
    <ld-menuitem>
      <mock:shadow-root>
        <li class="ld-menuitem" role="menuitem">
          <ld-button class="ld-menuitem__button" justifycontent="start" ldtabindex="-1" mode="neutral-ghost" size="sm" type="button">
            <slot></slot>
          </ld-button>
        </li>
      </mock:shadow-root>
      Menu item 3
    </ld-menuitem>
  </ld-menuitem-group>
</ld-menu>
`;

exports[`ld-menu renders 1`] = `
<ld-menu>
  <mock:shadow-root>
    <ul class="ld-menu" part="list" role="menu">
      <slot></slot>
    </ul>
  </mock:shadow-root>
  <ld-menuitem>
    Menu item
  </ld-menuitem>
</ld-menu>
`;
